### ISSUE TITLE
Mouse dragging breaks in pangoterm

### ISSUE THREAD

**Issue by leonerd** [neovim/neovim#5583]
In `xterm`, both regular vim and neovim can perform mouse-drag operations correctly.

In [`pangoterm`][1], a regular vim works fine but neovim suffers a breakage, that after the first drag movement of the mouse, no further movement of the mouse will have an effect until the mouse button is released.

[1]: https://launchpad.net/pangoterm

**Comment by leonerd** [comment]
I have been debugging this by capturing bytes read/written by nvim. What I discover is that neovim does:

```
((mouse click))
IN 9 bytes:
  1B 5B 3C 30 3B 35 3B 31 4D                      | .[<0;5;1M
OUT 164 bytes:
  1B 5B 3F 32 35 6C 1B 5B 31 3B 32 33 48 1B 5B 3F | .[?25l.[1;23H.[?
  31 32 6C 1B 5B 3F 32 35 68 1B 5B 3F 32 35 6C 1B | 12l.[?25h.[?25l.
  5B 31 3B 32 33 48 1B 5B 3F 31 32 6C 1B 5B 3F 32 | [1;23H.[?12l.[?2
  35 68 1B 5B 3F 32 35 6C 1B 5B 31 3B 32 33 48 1B | 5h.[?25l.[1;23H.
  5B 3F 31 32 6C 1B 5B 3F 32 35 68 1B 5B 3F 32 35 | [?12l.[?25h.[?25
  6C 1B 5B 31 3B 32 33 48 1B 5B 3F 31 32 6C 1B 5B | l.[1;23H.[?12l.[
  3F 32 35 68 1B 5B 3F 32 35 6C 1B 5B 32 33 3B 36 | ?25h.[?25l.[23;6
  35 48 35 20 1B 5B 31 3B 35 48 1B 5B 31 3B 35 48 | 5H5 .[1;5H.[1;5H
  1B 5B 3F 31 32 6C 1B 5B 3F 32 35 68 1B 5B 3F 32 | .[?12l.[?25h.[?2
  35 6C 1B 5B 31 3B 35 48 1B 5B 3F 31 32 6C 1B 5B | 5l.[1;5H.[?12l.[
  3F 32 35 68                                     | ?25h

((mouse drag))
IN 10 bytes:
  1B 5B 3C 33 32 3B 36 3B 31 4D                   | .[<32;6;1M
OUT 72 bytes:
  1B 5B 3F 32 35 6C 1B 5B 31 3B 35 48 1B 5B 3F 31 | .[?25l.[1;5H.[?1
  32 6C 1B 5B 3F 32 35 68 1B 5B 3F 32 35 6C 1B 5B | 2l.[?25h.[?25l.[
  31 3B 35 48 1B 5B 3F 31 32 6C 1B 5B 3F 32 35 68 | 1;5H.[?12l.[?25h
  1B 5B 3F 32 35 6C 1B 5B 31 3B 35 48 1B 5B 3F 31 | .[?25l.[1;5H.[?1
  32 6C 1B 5B 3F 32 35 68                         | 2l.[?25h
OUT 61 bytes:
  1B 5B 3F 32 35 6C 1B 5B 3F 31 30 30 32 68 1B 5B | .[?25l.[?1002h.[
  3F 31 30 30 36 68 1B 5B 3F 31 30 30 32 68 1B 5B | ?1006h.[?1002h.[
  3F 31 30 30 36 68 1B 5B 32 20 71 1B 5B 31 3B 35 | ?1006h.[2 q.[1;5
  48 1B 5B 3F 31 32 6C 1B 5B 3F 32 35 68          | H.[?12l.[?25h

((no more mouse events are received...))
```

Whereas regular vim does

```
((mouse click))
IN 6 bytes:
  1B 5B 4D 20 25 21                               | .[M %!
OUT 16 bytes:
  1B 5B 32 33 3B 36 35 48 35 20 1B 5B 31 3B 35 48 | .[23;65H5 .[1;5H

((mouse drag))
IN 6 bytes:
  1B 5B 4D 40 26 21                               | .[M@&!
OUT 60 bytes:
  1B 5B 31 20 71 1B 5B 3F 32 35 6C 1B 5B 6D 1B 5B | .[1 q.[?25l.[m.[
  32 34 3B 34 48 1B 5B 31 6D 28 69 6E 73 65 72 74 | 24;4H.[1m(insert
  29 2D 08 20 56 49 53 55 41 4C 20 2D 2D 1B 5B 6D | )-. VISUAL --.[m
  1B 5B 34 38 43 32 1B 5B 31 3B 36 48             | .[48C2.[1;6H
OUT 46 bytes:
  08 1B 5B 34 38 3B 35 3B 32 34 32 6D 74 1B 5B 6D | ..[48;5;242mt.[m
  1B 5B 32 33 3B 36 35 48 1B 5B 31 6D 1B 5B 37 6D | .[23;65H.[1m.[7m
  36 20 1B 5B 31 3B 36 48 1B 5B 3F 32 35 68       | 6 .[1;6H.[?25h
IN 6 bytes:

((more drag...))
  1B 5B 4D 40 27 21                               | .[M@'!
OUT 70 bytes:
  1B 5B 3F 32 35 6C 1B 5B 6D 1B 5B 32 34 3B 37 30 | .[?25l.[m.[24;70
  48 33 1B 5B 31 3B 37 48 08 1B 5B 34 38 3B 35 3B | H3.[1;7H..[48;5;
  32 34 32 6D 68 1B 5B 6D 1B 5B 32 33 3B 36 35 48 | 242mh.[m.[23;65H
  1B 5B 31 6D 1B 5B 37 6D 37 20 1B 5B 31 3B 37 48 | .[1m.[7m7 .[1;7H
  1B 5B 3F 32 35 68                               | .[?25h
IN 6 bytes:
  1B 5B 4D 40 28 21                               | .[M@(!
```

I believe what is happening here is that after it receives a drag event, neovim is resetting the DEC modes 1002 and 1006 (which relate to mouse reporting), which is upsetting the `libvterm` underlying `pangoterm`. It is possible this is also a libvterm bug in that it ought not get upset by such a request, but also neovim ought not be fiddling with the mode during an ongoing drag operation anyway.

**Comment by leonerd** [comment]
This turns out in part to indeed be a libvterm bug (or at least issue - it's unclear from terminal specs whether resetting the dragging mode while a drag operation is in progress should break, but I feel it useful that it doesn't) - https://bugs.launchpad.net/libvterm/+bug/1640917

That's now fixed; but it would still be nice for neovim not to make unnecessary idempotent terminal mode changes anyway; namely the line

```
OUT 61 bytes:
  1B 5B 3F 32 35 6C 1B 5B 3F 31 30 30 32 68 1B 5B | .[?25l.[?1002h.[
  3F 31 30 30 36 68 1B 5B 3F 31 30 30 32 68 1B 5B | ?1006h.[?1002h.[
  3F 31 30 30 36 68 1B 5B 32 20 71 1B 5B 31 3B 35 | ?1006h.[2 q.[1;5
  48 1B 5B 3F 31 32 6C 1B 5B 3F 32 35 68          | H.[?12l.[?25h
````

**Comment by justinmk** [comment]
@leonerd Thanks for debugging that. We're sending that request every time `setmouse()` is called. We could instead keep an internal flag when we want to "enable/disable mouse".

**Comment by leonerd** [comment]
@justinmk - I'm quite keen to start getting into working on the terminal IO layer of nvim, so if you want to assign that to me and give me a couple of pointers to code, it might make a good one for me to start on

**Comment by justinmk** [comment]
@leonerd Wonderful! Here are some general notes:

- The terminal UI is implemented in `tui/tui.c` and `tui/input.c`. That's why e.g. [#5556](https://github.com/neovim/neovim/pull/5556/files) changes both files to make a request and later respond to it (except the "response" is not implemented in #5556, because that's the part that requires work...). 
- The terminal UI is decoupled from the "logical" UI (unlike Vim, where the terminal UI is _the_ UI).
    - Compare `screen_char()` in Vim: it eventually reduces to `ui_write()`, which just calls `write(2)`.
    - ...vs `screen_char()` in Nvim: it calls `ui_puts()`, which does a `UI_CALL(put)`.
- `UI_CALL` is the abstraction that sends UI data to different UIs. The built-in TUI is itself "just another UI", it is completely decoupled--in fact it runs on _its own thread_.

The work for handling a request/response termcode is essentially: 

1. capturing the response in `tui/input.c` (easy) 
2. communicate the response back to `tui/tui.c` and deal with it.
    - This involves (probably) adding new handlers [here](https://github.com/neovim/neovim/blob/36c0ec6dd49c8c1a57eaa0b9f9d3c44792582f37/src/nvim/tui/tui.c#L83-L114)
    - What isn't clear to me yet is how to deal with the "race": Should the TUI startup wait N milliseconds for a response before starting up (`terminfo_start()`)? Or should we happily continue, then teardown and re-init the TUI whenever a response happens to be received?
